diff --git a/threads/init.c b/threads/init.c
index aae924d..27abedd 100644
--- a/threads/init.c
+++ b/threads/init.c
@@ -165,6 +165,18 @@ paging_init (uint64_t mem_end) {
 
 	// reload cr3
 	pml4_activate(0);
+       // Enforcing Read-only Page Write Protection for Ring 0 Code.
+       // See the AMD64 System Architecture Programmer's Manual, Volume 2, Section 5.6.4
+       //
+       // # Safety
+       // Interrupt must be disabled from getting CR0 register to applying the register's value.
+
+       uint64_t enforce_wp_kern = 1 << 16;
+       uint64_t cr0;
+
+       __asm __volatile("movq %%cr0, %0" : "=r"(cr0));
+       cr0 |= enforce_wp_kern;
+       __asm __volatile("movq %0, %%cr0" : : "r"(cr0));
 }
 
 /* Breaks the kernel command line into words and returns them as
